packages/server-side-render/src/server-side-render.js

```diff
@@ -20,126 +20,105 @@ export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
 			urlQueryArgs = {},
-		} = props;
+		} = options;
 
 		// If httpMethod is 'POST', send the attributes in the request body instead of the URL.
 		// This allows sending a larger attributes object than in a GET request, where the attributes are in the URL.
 		const isPostRequest = 'POST' === httpMethod;
 		const urlAttributes = isPostRequest ? null : attributes;
 		const path = rendererPath( block, urlAttributes, urlQueryArgs );
 		const data = isPostRequest ? { attributes } : null;
+		let currentFetchRequest = null;
 
 		// Store the latest fetch request so that when we process it, we can
 		// check if it is the current request, to avoid race conditions on slow networks.
-		const fetchRequest = ( this.currentFetchRequest = apiFetch( {
+		const fetchRequest = ( currentFetchRequest = apiFetch( {
 			path,
 			data,
 			method: isPostRequest ? 'POST' : 'GET',
 		} )
-			.then( ( response ) => {
+			.then( ( fetchResponse ) => {
 				if (
-					this.isStillMounted &&
-					fetchRequest === this.currentFetchRequest &&
-					response
+					isMounted &&
+					fetchRequest === currentFetchRequest &&
+					fetchResponse
 				) {
-					this.setState( { response: response.rendered } );
+					setResponse( fetchResponse.rendered );
+					setIsLoading( false );
 				}
 			} )
 			.catch( ( error ) => {
-				if (
-					this.isStillMounted &&
-					fetchRequest === this.currentFetchRequest
-				) {
-					this.setState( {
-						response: {
-							error: true,
-							errorMsg: error.message,
-						},
+				if ( isMounted && fetchRequest === currentFetchRequest ) {
+					setResponse( {
+						error: true,
+						errorMsg: error.message,
 					} );
+					setIsLoading( false );
 				}
 			} ) );
 		return fetchRequest;
 	}
 
-	render() {
-		const response = this.state.response;
-		const {
-			className,
-			EmptyResponsePlaceholder,
-			ErrorResponsePlaceholder,
-			LoadingResponsePlaceholder,
-		} = this.props;
-
-		if ( response === '' ) {
-			return (
-				<EmptyResponsePlaceholder
-					response={ response }
-					{ ...this.props }
-				/>
-			);
-		} else if ( ! response ) {
-			return (
-				<LoadingResponsePlaceholder
-					response={ response }
-					{ ...this.props }
-				/>
-			);
-		} else if ( response.error ) {
-			return (
-				<ErrorResponsePlaceholder
-					response={ response }
-					{ ...this.props }
-				/>
-			);
-		}
-
+	if ( response === '' ) {
+		return <EmptyResponsePlaceholder response={ response } { ...props } />;
+	}
+	if ( isLoading ) {
 		return (
-			<RawHTML key="html" className={ className }>
-				{ response }
-			</RawHTML>
+			<LoadingResponsePlaceholder>
+				<RawHTML key="html" className={ className }>
+					{ response }
+				</RawHTML>
+			</LoadingResponsePlaceholder>
 		);
 	}
+	if ( response?.error ) {
+		return <ErrorResponsePlaceholder response={ response } { ...props } />;
+	}
+
+	return (
+		<RawHTML key="html" className={ className }>
+			{ response }
+		</RawHTML>
+	);
 }
 
 ServerSideRender.defaultProps = {
```

Comment:
[`defaultProps` may disappear from React in the future](https://github.com/reactjs/rfcs/pull/107), and it's not necessary for function components anyway since those can just use default arguments. So I'd recommend not using it.

Reply:
@ZebulanStanphill agreed and done  👍